### PR TITLE
Mitigate shape mismatch issue for optimize_ddp_lazy_compile=True

### PR DESCRIPTION
**Pull request by BoyuanFeng**
Summary:
`DDP Optimizer lazy compile` does not support dynamic shape yet and may report shape mismatch error. See detail: https://github.com/pytorch/pytorch/issues/116300

The root cause is that, in lazy compile, compilation is not conducted until the first `forward` call, while the symbolic shape  information is missing at that time.

This diff detects dynamic shapes and explicitly raise an user error.

Test Plan: CI

Differential Revision: D54091447

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

### CODE REVIEW

**Review by Skylion007 — APPROVED** [review]

File: test/distributed/test_dynamo_distributed.py [hunk 1]
```diff
@@ -724,6 +724,7 @@ def opt_fn(inputs):
 
 
     @patch.object(config, "optimize_ddp", True)
+    @patch.object(config, "optimize_ddp_lazy_compile", True)
```
Comment (yf225) [inline test/distributed/test_dynamo_distributed.py:727]: wonder is there a way to check that it does produce the error message that you added in this PR?

File: test/distributed/test_dynamo_distributed.py [hunk 2]
```diff
@@ -749,14 +750,20 @@ def forward(self, x):
...
 
-        x_2 = torch.randn((B * 3, N, D), dtype=torch.float32, device=self.device)
-        self.assertTrue(same(mod(x_2), ddp_compiled_mod(x_2)))
+            x_2 = torch.randn((B * 3, N, D), dtype=torch.float32, device=self.device)
+            self.assertTrue(same(mod(x_2), ddp_compiled_mod(x_2)))
+        self.assertTrue(any("When turning on both config.ddp_optimizer and config.optimize_ddp_lazy_compile, "
+                            "we expect input tensor shapes remain static but found dynamic shapes instead. "
+                            "Please try to a) pad the input tensor to a static shape, or b) set "
+                            "config.ddp_optimizer=False. Note that config.ddp_optimizer=False may slowdown "
+                            "the performance" in msg for msg in cm.output))
```
Comment (yf225) [inline test/distributed/test_dynamo_distributed.py:766]: maybe option (c) is to try to turn off `config.optimize_ddp_lazy_compile`?

**Review by eellison — COMMENTED** [review]
I think we should land https://github.com/pytorch/pytorch/pull/120523 instead of this PR. We can support this case without throwing.

**Comment by eellison** [comment]
Hmm, maybe we can land this, but fundamentally there's nothing preventing us from using the strides that inductor computes from one layer of DDP Optimizer to the next. Ideally we wouldn't have a matrix of supported/unsupported different options.

**Pull request closed without merging**